mlir/tools/mlir-clang/Lib/clang-mlir.cc

```diff
@@ -231,48 +313,94 @@ bool MLIRScanner::getConstantStep(clang::ForStmt *fors, int64_t &step) {
 // for (int i = constant; i < other_constant; i++ or ++i)
 bool MLIRScanner::isTrivialAffineLoop(clang::ForStmt *fors,
                                       AffineLoopDescriptor &descr) {
-  if (!getConstantLowerBound(fors, descr.lowerBound, descr.indVar,
-                             descr.indVarType))
+  bool forwardloop;
+  if (!getConstantStep(fors, descr, forwardloop)) {
+    LLVM_DEBUG(dbgs() << "getConstantStep -> false\n");
     return false;
-  if (!getConstantUpperBound(fors, descr.upperBound, descr.indVar))
+  }
+  if (!getLowerBound(fors, descr, forwardloop)) {
+    LLVM_DEBUG(dbgs() << "getLowerBound -> false\n");
     return false;
-  if (!getConstantStep(fors, descr.step))
+  }
+  if (!getUpperBound(fors, descr, forwardloop)) {
+    LLVM_DEBUG(dbgs() << "getUpperBound -> false\n");
     return false;
+  }
   return true;
 }
 
+void MLIRScanner::buildAffineLoopImpl(clang::ForStmt *fors, mlir::Location loc,
+                                      mlir::Value lb, mlir::Value ub,
+                                      const AffineLoopDescriptor &descr) {
+  buildAffineLoopNest(
+      builder, loc, lb, ub, descr.getStep(),
+      [&](OpBuilder &nestedBuilder, mlir::Location loc, ValueRange ivs) {
+        SmallVector<mlir::Value, 1> iv(ivs);
+        assert(ivs.size() == 1 && "expect single ind var");
+        mlir::Value val = iv[0];
+        if (!descr.getForwardMode()) {
+          val = nestedBuilder.create<mlir::SubIOp>(loc, val, lb);
+          val = nestedBuilder.create<mlir::SubIOp>(
+              loc,
+              nestedBuilder.create<mlir::SubIOp>(loc, ub, getConstantIndex(1)),
+              val);
+        }
+        auto idx =
+            nestedBuilder.create<mlir::IndexCastOp>(loc, val, descr.getType());
+        createAndSetAllocOp(descr.getName(), idx, 0);
+
+        auto oldpoint = builder.getInsertionPoint();
+        auto oldblock = builder.getInsertionBlock();
+
+        builder.setInsertionPoint(nestedBuilder.getInsertionBlock(),
+                                  nestedBuilder.getInsertionPoint());
+
+        // TODO: set loop context.
+        Visit(fors->getBody());
+        // auto endBlock = builder.getInsertionBlock();
+        // if (endBlock->empty() || endBlock->back().isKnownNonTerminator()) {
+        //  builder.create<AffineYieldOp>(loc);
+        //}
+
+        nestedBuilder.setInsertionPoint(builder.getInsertionBlock(),
+                                        builder.getInsertionPoint());
+
+        builder.setInsertionPoint(oldblock, oldpoint);
```

Comment:
TODO: set the value of the iteration value to the final bound at the end of the loop